assets/js/workflow-reducer.js

```diff
@@ -25,312 +41,567 @@ export function mockAPI(mock_api) {
-  return (
-    api.loadWorkflow(getPageID())
-    .then(json => ({  type: RELOAD_WORKFLOW, workflow: json }))
-  )
-}
+// -- Workflow --
+
 
+// INITIAL_LOAD_WORKFLOW
+// Load the workflow for the first time
 export function initialLoadWorkflowAction() {
-  return (
-    api.loadWorkflow(getPageID())
-    .then(json => ({ type: INITIAL_LOAD_WORKFLOW, workflow: json }))
-  )
+  return {
+    type: INITIAL_LOAD_WORKFLOW,
+    payload: api.loadWorkflow(getPageID()).then((json) => {return json})
+  }
 }
+registerReducerFunc(INITIAL_LOAD_WORKFLOW + '_FULFILLED', (state, action) => {
+  // Sets the selected module from backend if it exists, or the first module if there are any at all
+  console.log("INITIAL_LOAD_WORKFLOW");
+  let selectedWfModule = null;
+
+  if (action.payload.selected_wf_module) {
+    selectedWfModule = action.payload.selected_wf_module;
+  } else if (action.payload.wf_modules && action.payload.wf_modules.length) {
+    selectedWfModule = action.payload.wf_modules[0].id;
+  }
 
-// Make an addModule call, then reload the workflow
+  return Object.assign({}, state, {
+    selected_wf_module: selectedWfModule,
+    workflow: action.payload
+  });
+});
+
+// RELOAD_WORKFLOW
+// Re-load the workflow
+// TODO: Do we need both initial and reload?
+export function reloadWorkflowAction() {
+  return {
+    type: RELOAD_WORKFLOW,
+    payload: api.loadWorkflow(getPageID()).then((json) => {return json})
+  }
+}
+registerReducerFunc(RELOAD_WORKFLOW + '_FULFILLED', (state, action) => {
+  console.log("RELOAD_WORKFLOW");
+  console.log("new workflow revision " + action.payload.revision);
+  /*var selectedWfModule = null;
+  if (action.payload.selected_wf_module) {
+    selectedWfModule = action.payload.selected_wf_module;
+  } else if (action.payload.wf_modules && action.payload.wf_modules.length) {
+    selectedWfModule = action.payload.wf_modules[0].id;
+  };*/
+  return update(state, {
+    //selected_wf_module: {$set: selectedWfModule},
+    workflow: {$merge: action.payload}
+  });
+});
+
+
+// ADD_MODULE
+// Add a module, then save the new module as the selected workflow
 export function addModuleAction(moduleId, insertBefore) {
-  return (
-    api.addModule(getPageID(), moduleId, insertBefore)
-    .then( data => store.dispatch(changeSelectedWfModuleAction(data.id)))
-    .then( reloadWorkflowAction )
-  )
+  return function (dispatch) {
+    return (
+
+      dispatch({
+        type: ADD_MODULE,
+        payload: api.addModule(getPageID(), moduleId, insertBefore)
+      }).then( ({value, action}) => {
+
+      dispatch(
+        setSelectedWfModuleAction(value.id)
+      );
+
+    }));
+  }
 }
+registerReducerFunc(ADD_MODULE + '_FULFILLED', (state, action) => {
+  let insertBefore = action.payload.insert_before;
 
-// Call delete API, then dispatch a reload
-export function removeModuleAction(id_to_delete) {
+  delete action.payload.insert_before;
+
+  return update(state, {
+    selected_wf_module: {$set: action.payload.id},
```

Comment:
Is this necessary, since you also dispatch setSelectedWfModuleAction? Maybe you can avoid that setSelectedWfModuleAction entirely, and hence the thunk.

Reply:
It's necessary to make the UI feel fast, yes, but it's not technically necessary. Without it there's a perceptible lag.

Reply:
Is there some way to dispatch both at once? I see no technical problem having two promises unresolved at the same time, and the server-side race condition is irrelevant here

Reply:
No, there isn't. You can't dispatch setSelectedWfModuleAction until you have a wfModuleId, which you only have once the wfModule returns from the database.